Design/name-mangling.md

```diff
@@ -78,111 +71,55 @@ We can identify the possible local names in this example:
-
-// foo: foo, fun_foo, fun_foo_takes_ABC, fun_foo_takes_ABC_returns_XYZ
-// x: x
-// B: B
-fun foo(x: Int, B: Int) { ... }
-
-// foo: foo, fun_foo, fun_foo_takes_Unit, fun_foo_takes_Unit_returns_Unit
-fun foo() {
-    // x: x
-    val x: Int = ...
-}
-```
-
-We start by resolving global names, which are necessary for
-namespace names.  The global entities are `class A`, `class
-B`, `fun A`, `fun foo(...)`, and `fun foo()`; for each, we
-pick the shortest name that isn't claimed by anything else.
-We end up with the following selection:
-* `class A` -> `class_A`
-* `class B` -> `B`
-* `fun A` -> `fun_A`
-* `fun foo(...)` -> `fun_foo_takes_x_B`
-* `fun foo()` -> `fun_foo_takes_Unit`
-
-This lets us resolve all namespace names that depend on
-names in the program, namely for `class A` and `class B`.
-(Note that while `fun foo()` is a namespace, the name
-of that namespace is `local`.)
-
-We then resolve the local names per namespace.  The
-interesting case here is in `class B`, where the getter and
-setter of `x` collide.
-
-Every namespace-name pair is now unique, so we get the
-following code, with unique names.  (Note that we compile
-properties to fields or methods; we do so here for B.x to
-demonstrate.)
-```kotlin
-class class_A(val class_A_x: Int) {
-    fun class_A_foo(): Int = ...
-}
-
-class B {
-    fun B_prop_x_getter(): Int { ... }
-    fun B_prop_x_setter(param_v: Int) { ... }
-}
-
-fun fun_A() { ... }
-
-fun fun_foo_takes_x_B(param_x: Int, param_B: Int) { ... }
-
-fun fun_foo_takes_Unit() {
-    val param_x: Int = ...
-}
-```
-
-We could run the algorithm again on namespace-name pairs,
-and this may be worth doing to remove bulky namespace names
-when they aren't necessary.  However, more thought is needed
-to see when that will make life easier.
+## Algorithm
+
+We maintain a **dependence graph** G over names. There is a 
+directed edge $x \to y$ iff determining the string representation 
+of $x$ requires the string representation of $y$.  
+$G$ is a DAG (see **Constraints**).
+
+### Initialization 
+1. For each name ```x```, build its candidate list.
+2. Build ```G```. 
+3. Compute initial strings for all names in a topological order of ```G```.
+### Main Loop
+Repeat until there are no conflicts: 
+1. **Detect conflicts**  
+Group names by ther current string and define 
+$$\text{Conflicted} := \{ x| \exists y \neq x:
+currentCandidate(x) = currentCandidate(y)\}$$.  
+If ```Conflicted``` is empty, we are done.
+2. **Choose names to fix**  
+Invoke  
+```FixSet = chooseNamesForFix(Conflicted)```
+where $\text{FixSet} \subseteq \text{Conflicted}, \text{FixSet}
+\neq \emptyset$
+3. **Advance candidates**  
+For each $x \in \text{FixSet}$, call ```deleteCurrentCandidate(x)```
+to move to the next candidate (or fallback).
+4. **Recompute affected string**  
+  For every name $v$ such that there exists a directed path $v\to x$
+in $G$ for some $x \in \text{FixSet}$, recompute
+```currentCandidate(v)``` in topological order.
```

Comment:
It isn't necessary to recompute this: candidates are symbolic, and rely on the names.  You can use a memoizing system to keep most of the order-related stuff implicit.